File: partiql-ast/src/main/java/org/partiql/ast/v1/expr/ExprCall.java
package org.partiql.ast.v1.expr;

import lombok.Builder;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.partiql.ast.v1.AstNode;
import org.partiql.ast.v1.AstVisitor;
import org.partiql.ast.v1.IdentifierChain;
import org.partiql.ast.v1.SetQuantifier;

import java.util.ArrayList;
import java.util.Collection;
import java.util.List;

/**
 * TODO docs, equals, hashcode
 */
@Builder
public class ExprCall extends Expr {
    @NotNull
    public final IdentifierChain function;

    @NotNull
    public final List<Expr> args;

    @Nullable
    public final SetQuantifier setq;

    public ExprCall(@NotNull IdentifierChain function, @NotNull List<Expr> args, @Nullable SetQuantifier setq) {
        this.function = function;
        this.args = args;
        this.setq = setq;
    }

    @NotNull
    @Override
    public Collection<AstNode> children() {
        List<AstNode> kids = new ArrayList<>();
        kids.add(function);
        kids.addAll(args);
        return kids;
    }

    @Override
    public <R, C> R accept(@NotNull AstVisitor<R, C> visitor, C ctx) {
        return visitor.visitExprCall(this, ctx);
    }
}

File: partiql-ast/src/main/kotlin/org/partiql/ast/v1/expr/ExprCall.kt
package org.partiql.ast.v1.expr

import org.partiql.ast.v1.AstNode
import org.partiql.ast.v1.AstVisitor
import org.partiql.ast.v1.Identifier
import org.partiql.ast.v1.SetQuantifier

/**
 * TODO docs, equals, hashcode
 */
public class ExprCall(
    @JvmField
    public var function: Identifier,
    @JvmField
    public var args: List<Expr>,
    @JvmField
    public var setq: SetQuantifier?,
) : Expr() {
    public override fun children(): Collection<AstNode> {
        val kids = mutableListOf<AstNode?>()
        kids.add(function)
        kids.addAll(args)
        return kids.filterNotNull()
    }

    public override fun <R, C> accept(visitor: AstVisitor<R, C>, ctx: C): R =
        visitor.visitExprCall(this, ctx)
}
